src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/vlboptionsfinder/VideoLinkBookingOptionsFinder.kt

```diff
@@ -0,0 +1,158 @@
+
+  /*
+    The periods of time when there are no appointments.
+    The keys are the times at which an empty period started
+    while the values are the corresponding times when the empty period ended.
+   */
+  private val emptyPeriods = TreeMap<LocalTime, LocalTime>()
+
+  init {
+    // The number of bookings in this room.
+    var currentBookings = 0
+
+    // The time of the previous event
+    var previousEventTime: LocalTime = LocalTime.MIN
+
+    // The time when the room last became unoccupied.
+    // This will be null when the room is currently occupied.
+    var freePeriodStarted: LocalTime? = LocalTime.MIN
+
+    events
+      .plus(StartEvent(LocalTime.MAX))
+      .sortedBy { it.time }
+      .forEach {
+        if (it.time > previousEventTime) {
+          /**
+           * More than one event can occur at a given time:
+           * One booking may end at the same time that another begins or there may be overlapping bookings.
+           *
+           * If the time of the current event is greater than the time of the previous event then
+           * all events for that earlier time have been seen and we can now make decisions about
+           * that previousEventTime (not the time of the current event):
+           */
+          if (currentBookings > 0) {
+            // The room became occupied or continued to be occupied at previousEventTime
+            if (freePeriodStarted != null) {
+              // The room became occupied at previousEventTime.  Record the empty period that ended at previousEventTime
+              emptyPeriods[freePeriodStarted!!] = previousEventTime
+              freePeriodStarted = null
+            }
+          } else {
+            // The room became unoccupied or continued to be unoccupied
+            if (freePeriodStarted == null) {
+              // The room became unoccupied. Remember the time at which this happened
+              freePeriodStarted = previousEventTime
+            }
+          }
+        }
+        // Now update state from the current event.
+        currentBookings += when (it) {
+          is StartEvent -> 1
+          is EndEvent -> -1
+        }
+        previousEventTime = it.time
+      }
+
+    // Finally, handle the end of the day
+    if (currentBookings > 0) {
+      if (freePeriodStarted != null) {
+        emptyPeriods[freePeriodStarted!!] = previousEventTime
+        freePeriodStarted = null
+      }
+    }
+  }
+
+  /**
+   * A List of pairs of start and end times representing the free periods.  Ordered by start time ascending.
+   * Mainly used for testing
+   */
+  fun emptyPeriods() = emptyPeriods.navigableKeySet().map {
+    Pair(it, emptyPeriods[it])
+  }
+
+  /**
+   * Answer the question 'is this Timeline free of bookings during the specified interval'.
+   */
+  fun isFreeForInterval(interval: Interval): Boolean {
+    if (interval.end.isBefore(interval.start)) {
+      throw IllegalArgumentException("start must precede end")
```

Comment:
Would this be better pushed up to `Interval`?

Reply:
Hmm.  It's a good check to have, but strictly speaking its not needed because the Intervals come from the incoming spec and that has validation it that would rule this out.  You're right though, I could add an isValid method to Interval.  Its already there on the IntervalValidator in the same file as Validator.